### ISSUE TITLE
Presubmit/Postsubmit/Sample test flakiness and long running time

### ISSUE THREAD

**Issue by jingzhang36** [kubeflow/pipelines#3256]

**Comment by numerology** [comment]
Among the four tests we currently have, e2e and sample tests are the major source of flakiness. The easiest treatment could be that we separate out those samples with external dependencies first.

**Comment by numerology** [comment]
Here's my proposal:

- Move XGBoost/Dataflow/AI platform/Kubeflow TF serving/(perhaps) container build samples into a new category, namely `integration tests`.
- Either mark `integration tests` optional or move them to Postsubmit. For changes that involve components, reviewers shall make sure integration tests are passed.
- As for the testing time, I think the current bottleneck is the image building (IIUC) + cluster creation + deployment. The test running time is tolerable per se. Perhaps we can take a look at how to optimize our docker image spec?

**Comment by Bobgy** [comment]
Thanks @numerology! That's a good plan.

I think moving sample tests with external dependency to post submit (and optional for presubmit) sounds great to me, sample-test is so far the longest running test ~50 min, e2e is ~40 min. Others are ignorable.

Some other ideas (haven't examined fully):
* Can we change some of the tests to run in parallel? (e.g. pipelines shouldn't interfere with each other, can we run all samples together?)

**Comment by Bobgy** [comment]
@numerology Because you didn't cover it here, do you plan to also target flakiness in this issue?
I think flakiness has very different fixes from long running time, maybe we can have a separate issue tracking that.

Do you want to own that too? I want to co-own flakiness part.

**Comment by numerology** [comment]
> Can we change some of the tests to run in parallel?

IIUC it's already running in parallel, see [here](https://github.com/kubeflow/pipelines/blob/d9bb1be25b838142267acbf83ca8f1a5bf33ef54/test/sample_test.yaml#L65). I think the bottleneck is some of the samples are significantly slower than others.

> I want to own or co-own flakiness part.

Yes we can collaborate on that part. Two things are on top of my head right now:

1. As explained in my first response, separate those test cases with external dependencies and make them post-submit or optional.

2. Apply [Argo retry](https://github.com/argoproj/argo/blob/master/examples/retry-with-steps.yaml#L18) in the test workflow. At least the image building and cluster deployment won't need to re-run, and most of the test cases are fast per se.

**Comment by numerology** [comment]
So I think the first part can be done in following PRs

- [ ] Prepare a separate entrypoint/script for the bespoken integration test
- [ ] Add that to prow config
- [ ] Remove corresponding cases in sample test
- [ ] Introduce retry to reduce flakiness.

**Comment by Bobgy** [comment]
> Prepare a separate entrypoint/script for the bespoken integration test

Because we have many tests/entry points, I feel like it will be helpful if we maintain a reused presubmit test entry point and postsubmit test entry point, and allow each test to be easily ported between presubmit/postsubmit.

What do you think?

**Comment by numerology** [comment]
> > Prepare a separate entrypoint/script for the bespoken integration test
> 
> Because we have many tests/entry points, I feel like it will be helpful if we maintain a reused presubmit test entry point and postsubmit test entry point, and allow each test to be easily ported between presubmit/postsubmit.
> 
> What do you think?

Agree. I think we can use a flag to control which set of cases will be executed.

**Comment by Bobgy** [comment]
visualization server image building seems really flaky recently, e.g. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/3403/kubeflow-pipeline-sample-test/1245123715558019072#1:build-log.txt%3A6224

because it downloads a lot of licenses from Github.

I think we should download and concatenate the licenses beforehand like other images.

**Comment by Bobgy** [comment]
* [ ] One more flakiness I see often recently: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/3483/kubeflow-pipeline-e2e-test/1248147971741061120#1:build-log.txt%3A199

**Comment by numerology** [comment]
> * [ ]  One more flakiness I see often recently: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/3483/kubeflow-pipeline-e2e-test/1248147971741061120#1:build-log.txt%3A199

Ack. It looks like to be some concurrency issue. Let me try to add some retry logic around it.

**Comment by Bobgy** [comment]
Adding the binding is actually the same for all tests. Maybe we can prepare a script that sets these up and no longer run any of these in actual tests.

**Comment by numerology** [comment]
> Adding the binding is actually the same for all tests. Maybe we can prepare a script that sets these up and no longer run any of these in actual tests.

Do you mean extract this common step out from upgrade/e2e/sample/integration tests? Curious how this can be done since each time we'll create a new cluster.

**Comment by Bobgy** [comment]
Failure is at this line: https://github.com/kubeflow/pipelines/blob/6b358d09a763d2b172c49cd389f34c5c683daa9b/manifests/kustomize/gcp-workload-identity-setup.sh#L111

In fact, because we always use the same KSAs and run in the same project. Workload identity binding changes on GSAs are the same (note that the command only refer to PROJECT, NAMESPACE, KSA, there is no cluster-name here.)

**Comment by Bobgy** [comment]
Fixed image building flakiness in https://github.com/kubeflow/pipelines/pull/3492

**Comment by Bobgy** [comment]
@numerology Do you have any other items left? I'm thinking most items are done. We can close this now.

If we identify any remaining ones, we can reopen.

**Comment by Bobgy** [comment]
Actually, postsubmit component test is still fairly flaky, but it isn't strictly required. We may only release on commits that passed them. WDUT?

**Comment by Bobgy** [comment]
/close
postsubmit test is the only problem now, but we can fix that later I think
